sqlalchemy_redshift/dialect.py

```diff
@@ -641,23 +643,85 @@ def keyfunc(column):
             'redshift_interleaved_sortkey': interleaved_sortkey,
         }
 
+    def _do_generate_iam_auth_token(self, redshift, cparams):
+        # Hostname is the cluster identifier
+        cluster_id = cparams['host']
+        response = redshift.describe_clusters(ClusterIdentifier=cluster_id)
+        # If there is no cluster, it will throw an exception
+        rs_cluster = response['Clusters'][0]
+
+        response = redshift.get_cluster_credentials(
+            DbUser=cparams['user'],
+            DbName=cparams['database'],
+            ClusterIdentifier=cluster_id,
+            AutoCreate=False)
+
+        # Extract user, password, hostname, and port from the response
+        return {
+            'host': rs_cluster['Endpoint']['Address'],
+            'port': rs_cluster['Endpoint']['Port'],
+            'user': response['DbUser'],
+            'password': response['DbPassword'],
+            'database': cparams['database']
+        }
+
+    def _generate_iam_auth_token(self, cparams):
+        """
+        Parses for necessary information to generate username and
+        password using the necessary Redshift API.
+
+        Username and database are passed to the get_cluster_credentials()
+        API to verify permissions.
+        """
+        try:
+            import boto3
+            if self.aws_region not in cparams:
+                raise ValueError(
+                    "Need AWS Region as part of the connection configuration.")
+
+            redshift = boto3.client('redshift',
+                                    region_name=cparams[self.aws_region])
+
+            # Verify that the parameters make sense.
+            if "password" in cparams:
+                raise ValueError(
+                    "'password' cannot be used when connecting with"
+                    " IAM federation."
+                )
+
+            return self._do_generate_iam_auth_token(redshift, cparams)
+
+        except ImportError:
+            raise ImportError(
+                "Could not find boto3 library, make sure the dependency "
+                "is installed."
+            )
+
     def create_connect_args(self, *args, **kwargs):
         """
         Build DB-API compatible connection arguments.
 
+        In addition, since Redshift offers different authentication
+        mechanisms, we use this method to dynamically retrieve
+        credentials via IAM, if the "iam" parameter is present.
+
         Overrides interface
         :meth:`~sqlalchemy.engine.interfaces.Dialect.create_connect_args`.
         """
         default_args = {
-            'sslmode': 'verify-full',
-            'sslrootcert': pkg_resources.resource_filename(
-                __name__,
-                'redshift-ca-bundle.crt'
-            ),
+            'sslmode':
+            'verify-full',
+            'sslrootcert':
+            pkg_resources.resource_filename(__name__,
```

Comment:
why did you move all this around?

Reply:
What are you using as auto-formatting settings so I don't have to manually adhere to the guide?

Reply:
no autoformat here at the moment. For multiline forms just make sure the closing bracket is on the same column as the start of the line that opens it, and every indent is a multiple of 4 and only 4 more than the previous line

```
# good
fn(
    example,
    example,
    example,
)

# good
fn(example, example, example)

# bad
fn(
    example,
    example,
    example
)

# bad
fn(example, example, example,)

# bad
fn(
    example,
    example,
    example)

# awful
fn(example,
   example,
   example)
```